### PR TITLE
Esri proxy

### PR DESCRIPTION
**Pull request by mki-c2c**
enable dpi awareness for layers from arcgis server by querying directly the REST API

### CODE REVIEW

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 1]
```diff
@@ -136,7 +137,69 @@ def internal_proxy_wms(self):
...
+                'map_resolution': 'dpi',
+                'transparent': 'transparent',
+                'bbox': 'bbox',
+            }
+            for param in self.request.params:
+                if param.lower() in param_dict:
+                    param_wms = param_wms + param_dict[param.lower()] + "=" + \
+                                urllib.parse.quote(
+                                self.request.params.get(param, '').encode('utf-8')
+                                ) + "&"
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/views/wms.py:159]: Dealing with query string «as a string» sounds a bit error prone to me.
What about storing them as dict, and finally `urlencode(` it at the end?

Comment (mki-c2c) [reply geoportal/geoportailv3_geoportal/views/wms.py:159]: good idea for improvement, I adapted the existing code without much change

Comment (mki-c2c) [reply geoportal/geoportailv3_geoportal/views/wms.py:159]: @tonio thanks for your review,
I just look at the code, the URL generation is manual everywhere, I wonder if I should not move this refactoring into a new PR (alongside with similar requests from PR #2547

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 2]
```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
             return True
         return False
 
+    def _get_arcgis_token(self, internal_wms, force_renew=False):
+        try:  # try to retrieve token from session storage
+            assert not force_renew
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:36]: Assert is disable on prod, use 

```
if not force_renew:
    raise AssertionError('blabla')
```

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 3]
```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
...
+            # try:
+            #     log.info(f"Check token at: {token_check_request.full_url}")
+            #     rep = urllib.request.urlopen(token_check_request, timeout=15)
+            #     check = json.load(rep)
+            #     if 'error' in check:
+            #         log.error(f"Token refused by: {token_check_request.full_url} - "
+            #                   f"server answered {check['error']}")
+            # except:
+            #     log.error(f"Failed token check at: {token_check_request.full_url}")
+            # assert 'error' not in check
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:55]: Should it be removed ?

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 4]
```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
...
+            token_data = {
+                'f': 'json',
+                'username': config["arcgis_token_username"],
+                'password': config["arcgis_token_password"],
+                'referer': 'x',
+                'expiration': 600
+            }
+            encoded_token_data = urllib.parse.urlencode(token_data).encode()
+            token_url = urllib.parse.urljoin(internal_wms.rest_url + '/', 'generateToken')
+            url_request = urllib.request.Request(token_url, data=encoded_token_data)
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:68]: Maybe using `requests` is more straight forward? (with `raise_on_status` and `json()`. Can also be mocked easily with https://pypi.org/project/responses/

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 5]
```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
...
         return False
 
+    def _get_arcgis_token(self, internal_wms, force_renew=False):
+        try:  # try to retrieve token from session storage
+            assert not force_renew
+            log.info('check token')
+            auth_token = self.request.session['auth_token']
+            log.info('token exists')
+            # renew token if it expires in the next 30s
+            assert datetime.fromtimestamp(float(auth_token['expires'])/1000 - 30) > datetime.now()
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:41]: Same

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 6]
```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
             return True
         return False
 
+    def _get_arcgis_token(self, internal_wms, force_renew=False):
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:34]: I find this function confusing. Maybe just doing the HTTP requests if the token is expired and not forced to be renewed ?

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 7]
```diff
@@ -136,7 +197,52 @@ def internal_proxy_wms(self):
...
                 return HTTPUnauthorized()
 
-        remote_host = internal_wms.url
+        query_params = {}
+        if internal_wms.ogc_server.type == 'arcgis':
+
+            # deactivated code for debug purpose
+            if False:
+                return Response('this is another hires layer\n',
+                                headers={"Content-Type": "text/plain"})
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:206]: :thinking: maybe this should be removed ?

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 8]
```diff
@@ -68,7 +130,6 @@ def _check_ip_for_httpsproxy(self, url):
 
     @view_config(route_name='wmspoi')
     def wmspoi(self):
-        config = self.request.registry.settings
         remote_host = os.environ["poi_server"]
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:133]: This is fragile. using `os.getenv()` with a default value seems more robust.

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 9]
```diff
@@ -136,7 +197,52 @@ def internal_proxy_wms(self):
...
+            }
+            for param, value in self.request.params.items():
+                lparam = param.lower()
+                if lparam in param_dict:
+                    query_params[param_dict[lparam]] = value
+                elif lparam == 'crs':
+                    if "EPSG:" in value:
+                        crs = value[5:]
+                    query_params["imageSR"] = crs
+                    query_params["bboxSR"] = crs
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:221]: `crs` could be used before assignment.

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 10]
```diff
@@ -149,32 +255,28 @@ def internal_proxy_wms(self):
             base64user = base64.b64encode(
                 "%s:%s" % (remote_user, remote_password)).replace("\n", "")
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:256]: Replacing this sentences with f-string could improve readability.

File: geoportal/geoportailv3_geoportal/views/wms.py [hunk 11]
```diff
@@ -149,32 +255,28 @@ def internal_proxy_wms(self):
...
         # TODO : Specific action when user is logged in ?
         # Forward authorization to the remote host
         # check sso
         if self.request.user and hasattr(self.request.user, 'ogc_role') and \
            self.request.user.ogc_role is not None and \
            self.request.user.ogc_role != -1:
-            param_wms += "roleOGC=%s&" % str(self.request.user.ogc_role)
+            query_params['roleOGC'] = self.request.user.ogc_role
 
         url = ""
```
Comment (RemiDesgrange) [inline geoportal/geoportailv3_geoportal/views/wms.py:277]: declare `url` here is useless.

**Review by RemiDesgrange — CHANGES_REQUESTED** [review]
I couldn't launch the docker composition without adding `ARCGIS_USER` and `ARCGIS_PASS` to the composition. Also. `http://localhost:8080/admin` doesn't work :-/

**Pull request merged**